docs/proposals/20220927-label-sync-between-machine-and-nodes.md

```diff
@@ -0,0 +1,131 @@
+
+### Non-Goals
+
+- Support for arbitrary/user-specified label prefixes (but the design doesn't prevent us to eventually reconsider this in future if there is strong demand for this feature)
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin/user, I would like a declarative and secure means by which to assign roles to my nodes. For example, when I do `kubectl get nodes`, I want the ROLES column to list my assigned roles for the nodes.
+
+#### Story 2
+
+As a cluster admin/user, for the purpose of workload placement, I would like a declarative and secure means by which I can add/remove labels on groups of nodes. For example, I want to run my outward facing nginx pods on a specific set of nodes (e.g. md-secure-zone) to reduce exposure to the company's publicly trusted server certificates.
+
+#### Story 3
+
+As a cluster admin/user, I want that Cluster API label management on Kubernetes Nodes do not conflits with labels directly managed by users or by other controllers.
+
+### Implementation Details/Notes/Constraints
+
+While designing a solution for syncing labels between Machine and underlying Kubernetes Nodes two main concerns have been considered:
+
+- Security, because Node labels can be used to attract workload to a specific Machine
+- Impacts on other components applying labels on Kubernetes Nodes, like e.g. Kubeadm, CPI etc.
+
+### Label Prefixes
+
+The idea of scoping syncronization to a well defined set of labels is a first answer to security/concurrency concerns; labels to be managed by Cluster API have been selected based on following criteria:
+
+- The `node-role.kubernetes.io` label has been used widely in the past to identify the role of aKubernetes Node (e.g. `node-role.kubernetes.io/worker=''`). For example, `kubectl get node` looks for this specific label when displaying the role to the user.
+
+- The `node-restriction.kubernetes.io/` prefix is recommended in the [Kubernetes docs](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction) for things such as workload placement; according to https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/ also
+  labels like `example.node-restriction.kubernetes.io/fips=true` falls in this category (prefix =~ domain).
+
+- Cluster API will define its own label prefix, `node.cluster.x-k8s.io/`; for consistency we consider also `example.node.cluster.x-k8s.io/gpu=true` in this category (prefix =~ domain).
+
+#### Synchronization of CAPI Labels
+
+Only the labels that fall under the prefix described above will need to be kept in sync between the Machine and Node objects.
+
+The syncronization process is going to use [server side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) in order to ensure that Cluster API will manage only the subset of labels coming from Machine objects and ignore labels applied concurrently by other users/other components.
```

Comment:
What happens during a move operation when the server side apply metadata is cleared?

Reply:
After restoring the controller will set again labels and metadata will be rebuilt, there is a gotcha in this operation (I have updated the next sequence to keep track of it; reporting here as well

> This requires to define an identity/manager name to be used by CAPI when performing this operation; additionally during implementation we are going to identify and address eventual additional steps required to properly transition existing Nodes to SSA, if required.
> From a preliminary investigation, the risk is that the manager that created the object and the new SSA manager will become co-owner of the same labels, and this will prevent deletion of labels.
Similar risks applies to clusterctl move or operations like velero backup/restore.

Reply:
I think it would be good for clusterctl move in general to not modif the managed fields. (I don't know what the current behavior is)

I think this can be problematic in general

Reply:
thinking back to move, this is not an issue for Nodes (they live in the workload cluster, so they are not involved in the operation)

Reply:
Very good catch. So not relevant for this proposal

Reply:
I think we can resolve